### PR TITLE
Feat/287 allocation of emissions

### PR DESCRIPTION
**Pull request by shon-button**
[Addresses](https://github.com/bcgov/cas-reporting/issues/287)

### 🚀 Impact:

Provides the functionality to allocate emissions to reported emission production data.

### ✏️ Notes
- This PR does not include tests.  Tests (and other potential tech debt) will be addressed in [Allocation of Emissions Tests](https://github.com/bcgov/cas-reporting/issues/431)

### 🔬 Local Testing:

1. From terminal command, start the api server:
 ```
cd bc_obps
make reset_db
make run
```  
2.  From terminal command, start the app development server:
 ```
cd bciers && yarn dev-all
```  

1. Navigate to http://localhost:3000
2. Click button "Log in with Business BCeID"
3. Navigate to http://localhost:3000/reporting/reports
4. On row `Operation 3`, click button `Start`
5. On page `Review operation information` , under 'Reporting activities', add `Carbonate use`.
6.  Click button `Save and Continue`
7. On page 'Person responsible', choose one of the options in 'Select contact if the are already a BCIERS user'.
8. Click button 'Save and Continue'.
9. On page 'Review facility information', click the check mark beside 'Carbonate use'.
10. Click button 'Save and Continue'.
11. Complete and save the activity forms: `General stationary combustion excluding line tracing` and 'Carbonate use'
*Note the 'Equivalent Emission' values after you have entered some test data*
13. Click button `Submit`
14. Change url to: http://localhost:3000/reporting/reports/3/facilities/f486f2fb-62ed-438d-bb3e-0819b51e3aff/production-data
15.  Complete the form: `Production Data`, making sure that at least 2 different products are selected.
16. Click button `Save and Continue`.
**Expected results:** 
  -  `Allocation of Emissions` page displays under step `Report Information`, task `Allocation of Emissions`
  - Category products display products reported on form: `Production Data`
  - Total emissions displays total as expected for categories: `Stationary fuel combustion emissions` and `Industrial process emissions`.
  - `Totals in tCO2e\Total emissions attributable for reporting` displays total as expected 
  - Button `Save and Continue` is disabled
  - Message displays `All emissions must be allocated to 100% before saving and continuing`
18. Complete form's required fields
19. Enter the category product allocated emissions values to match the category's total emissions
**Expected results:** 
  -  Category `Total Allocated` value reflects category product field(s) value(s) as a percent of the category `Total Emissions` value
  - `Totals in tCO2e` products matches category product field(s) value(s) 
   - Button `Save and Continue` is NOT disabled
  - Message does NOT display `All emissions must be allocated to 100% before saving and continuing`
 20. Click Button `Save and Continue`
**Expected results:** 
  - POST request is successful and page routes to ` Additional Reporting Data`
 21. Change url to: http://localhost:3000/reporting/reports/3/facilities/f486f2fb-62ed-438d-bb3e-0819b51e3aff/allocation-of-emissions
**Expected results:** 
  - Data reflects data entered in the allocation of emissions form
 22. Change the allocation of emissions by either changing the amounts allocated to different products or changing all allocations from one product to another.
 23. Click button 'Save & Continue'
 24.  Change url to: http://localhost:3000/reporting/reports/3/facilities/f486f2fb-62ed-438d-bb3e-0819b51e3aff/allocation-of-emissions
 **Expected results:**
    - Form data reflects changes made in step 22

### CODE REVIEW

File: bc_obps/reporting/service/report_emission_allocation_service.py
Comment (dleard) [inline bc_obps/reporting/service/report_emission_allocation_service.py:65]: Is this block necessary? Isn't the total for emissions attributable for reporting already calculated in the EmissionCategoryService?

File: bc_obps/reporting/service/report_emission_allocation_service.py
Comment (dleard) [inline bc_obps/reporting/service/report_emission_allocation_service.py:117]: We should probably carry the emission category Ids in the schema / form data as hidden fields & use them here rather than fetching by name. Fetching objects by name is fragile.

File: bciers/apps/reporting/src/data/jsonSchema/facility/facilityEmissionAllocation.tsx
Comment (dleard) [inline bciers/apps/reporting/src/data/jsonSchema/facility/facilityEmissionAllocation.tsx:113]: :heart: There may be other places in the app where this will be useful

**Review by dleard — APPROVED** [review]
Thanks for all the hard work on this! I've got a few things I found & I'll break them up by Needs fixing, Tech debt & Note. I'll approve this pending the check that the save buttons behave properly as I think that's the only thing that should hold this back for now.

**Needs Fixing:**
- The "Save'" and "Save and Continue" buttons do the same thing. Which bypasses the 100% allocation check. This may have been fixed in #2519. We'll need to make sure after rebasing this that the buttons behave as intended here.

**Tech Debt:**
- Records are being saved to the report_emission_allocation table in the database when the allocated value is 0. We shouldn't be saving 0 records (high priority tech debt)
- The comments I added to the code can be added as tech debt
- The details box when methodology is other is probably too small & the font in it is different than the rest of the page. Screenshot below
- The "Total Emissions" field at the bottom of the page is of inconsistent width with the rest of the page. Screenshot below

**Note:**
- Something to run by POs/designers/BA: The "Total emissions attributable for reporting" field at the bottom of the page does not match the totals attributed to products when there are excluded emissions attributed to products. Screenshot below.

Screenshots:

Description box small & different font:
Screenshot from 2024-12-09 11-30-29

Attributed to products totals & total attributable does not match when allocating excluded emissions + inconsistent width of field:
Screenshot from 2024-12-09 11-48-53

**Comment by shon-button** [comment]
> Thanks for all the hard work on this! I've got a few things I found & I'll break them up by Needs fixing, Tech debt & Note. I'll approve this pending the check that the save buttons behave properly as I think that's the only thing that should hold this back for now.
> 
> **Needs Fixing:**
> 
> * The "Save'" and "Save and Continue" buttons do the same thing. Which bypasses the 100% allocation check. This may have been fixed in [392 button tech debt #2519](https://github.com/bcgov/cas-registration/pull/2519). We'll need to make sure after rebasing this that the buttons behave as intended here.
> 
> **Tech Debt:**
> 
> * Records are being saved to the report_emission_allocation table in the database when the allocated value is 0. We shouldn't be saving 0 records (high priority tech debt)
> * The comments I added to the code can be added as tech debt
> * The details box when methodology is other is probably too small & the font in it is different than the rest of the page. Screenshot below
> * The "Total Emissions" field at the bottom of the page is of inconsistent width with the rest of the page. Screenshot below
> 
> **Note:**
> 
> * Something to run by POs/designers/BA: The "Total emissions attributable for reporting" field at the bottom of the page does not match the totals attributed to products when there are excluded emissions attributed to products. Screenshot below.
> 
> Screenshots:
> 
> Description box small & different font: Screenshot from 2024-12-09 11-30-29
> 
> Attributed to products totals & total attributable does not match when allocating excluded emissions + inconsistent width of field: Screenshot from 2024-12-09 11-48-53

@dleard 
Thanks very much for the review!

I believe the `Save` button appearance is new from the merged button PR...I'll start a new branch to resolve this.
I struggled with sizing the textarea field; so, any insight on RJSF configs for textareas much appreciated - otherwise I'll keep struggling to get it styled appropriately. :)

**Comment by acatchpole** [comment]
>Attributed to products totals & total attributable does not match when allocating excluded emissions + inconsistent width of field:
This bug is because non-attributable emissions are being included in the total allocations by report product that is calculated in the backend. The fix for this will be included in the ticket to finish up this work.

**Pull request merged**
